running usb_cam_node does not give proper feedback when using a video device without proper permissions

**Issue Description**
When attempting to run the usb_cam_node with a /dev/video<x> device without the necessary permissions, the node fails and crashes. The problem lies in the error handling, as the exception is not adequately handled, resulting in a vague error message.

**Current Behavior**
The current log output when starting the node without proper permission is as follows:
```
❯ ros2 run usb_cam usb_cam_node_exe
[INFO] [1698243801.072328729] [usb_cam]: camera_name value: default_cam
[WARN] [1698243801.072534313] [usb_cam]: framerate: 30.000000
[INFO] [1698243801.073781020] [usb_cam]: using default calibration URL
[INFO] [1698243801.073798354] [usb_cam]: camera calibration URL: file:///home/vm/.ros/camera_info/default_cam.yaml
[ERROR] [1698243801.073881770] [camera_calibration_parsers]: Unable to open camera calibration file [/home/vm/.ros/camera_info/default_cam.yaml]
[WARN] [1698243801.073921020] [usb_cam]: Camera calibration file /home/vm/.ros/camera_info/default_cam.yaml not found
[INFO] [1698243801.073936645] [usb_cam]: Starting 'default_cam' (/dev/video0) at 640x480 via mmap (yuyv) at 30 FPS
terminate called after throwing an instance of 'char*'
[ros2run]: Aborted
```

**Proposed Solution**
*Notice: I am not a skilled c++ Developer, so I might be wrong*
It appears that a potential solution could involve modifying the [usb_cam_node.cpp](https://github.com/ros-drivers/usb_cam/blob/ros2/src/usb_cam_node.cpp) file. Specifically, you might consider adding a try-catch block around the code segment between line 148 and 156. This try-catch block should provide a more informative error message for cases where proper permissions are lacking.

--- a/.github/workflows/build_test.yml
+++ b/.github/workflows/build_test.yml
@@ -1,11 +1,11 @@
 name: ROS2 CI
 
 on:
-  workflow_dispatch:
   pull_request:
     branches:
       - 'ros2/*'
       - 'ros2'
+  workflow_dispatch:
 
 jobs:
   test_environment:

--- a/include/usb_cam/utils.hpp
+++ b/include/usb_cam/utils.hpp
@@ -29,6 +29,10 @@
 #ifndef USB_CAM__UTILS_HPP_
 #define USB_CAM__UTILS_HPP_
 
+extern "C" {
+#include <fcntl.h>  // for open()
+}
+
 #include <sys/ioctl.h>
 #include <sys/time.h>
 #include <cmath>
@@ -36,6 +40,7 @@
 #include <cstring>
 #include <sstream>
 #include <string>
+#include <map>
 
 #include "linux/videodev2.h"
 
@@ -130,6 +135,47 @@ inline io_method_t io_method_from_string(const std::string & str)
   } else {
     return io_method_t::IO_METHOD_UNKNOWN;
   }
+}
+
+/// \brief List currently available valid V4L2 devices
+///   Can be used to check if a device string is valid before
+///   starting up.
+///
+///   Inspired by: http://stackoverflow.com/questions/4290834/how-to-get-a-list-of-video-capture-devices-web-cameras-on-linux-ubuntu-c
+inline std::map<std::string, v4l2_capability> available_devices()
+{
+  // Initialize vector of device strings to fill in
+  std::map<std::string, v4l2_capability> v4l2_devices;
+
+  // V4L2 spec says there can only be a maximum of 64 devices
+  const uint8_t MAX_DEVICES = 64;
+
+  for (uint8_t i = 0; i < MAX_DEVICES; i++) {
+    std::string device_str = "/dev/video" + std::to_string(i);
+    // See if device exists
+    if (access(device_str.c_str(), F_OK) != -1) {
+      int fd;
+      // Try and open device to test access
+      if ((fd = open(device_str.c_str(), O_RDONLY)) == -1) {
+        std::cerr << "Cannot open device: `" << device_str << "`, ";
+        std::cerr << "double-check read / write permissions for device" << std::endl;
+      } else {
+        struct v4l2_capability device_capabilities = {};
+
+        if (xioctl(fd, VIDIOC_QUERYCAP, &device_capabilities) == -1) {
+          std::cerr << "Could not retrieve device capabilities: `" << device_str;
+          std::cerr << "`" << std::endl;
+        } else {
+          v4l2_devices[device_str] = device_capabilities;
+        }
+      }
+    } else {
+      // device doesn't exist, break
+      break;
+    }
+  }
+
+  return v4l2_devices;
 }
 
 }  // namespace utils

--- a/src/usb_cam_node.cpp
+++ b/src/usb_cam_node.cpp
@@ -138,6 +138,23 @@ void UsbCamNode::init()
     m_camera_info->setCameraInfo(*m_camera_info_msg);
   }
 
+  // Check if given device name is an available v4l2 device
+  auto available_devices = usb_cam::utils::available_devices();
+  if (available_devices.find(m_parameters.device_name) == available_devices.end()) {
+    RCLCPP_ERROR_STREAM(
+      this->get_logger(),
+      "Device specified is not available or is not a vaild V4L2 device: `"
+        << m_parameters.device_name << "`"
+    );
+    RCLCPP_INFO(this->get_logger(), "Available V4L2 devices are:");
+    for (const auto & device : available_devices) {
+      RCLCPP_INFO_STREAM(this->get_logger(), "    " << device.first);
+      RCLCPP_INFO_STREAM(this->get_logger(), "        " << device.second.card);
+    }
+    rclcpp::shutdown();
+    return;
+  }
+
   m_image_msg->header.frame_id = m_parameters.frame_id;
   RCLCPP_INFO(
     this->get_logger(), "Starting '%s' (%s) at %dx%d via %s (%s) at %i FPS",

